lib/recursive-methods.rb

```diff
@@ -1,27 +1,56 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) - > the function will run as many times as the given number.
+# Space complexity: O(1) -> we're not creating new variables but the method is taking up space in the stack until the base case equals true, then the stacks will popped of 1 by 1 starting from the top of the stack.
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+    # raise NotImplementedError, "Method not implemented"
+    if n < 0
+      raise ArgumentError
+    elsif n == 1 || n == 0
+      return 1
+    end
+    return n * factorial(n - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+def fib(n)
+  if n < 2
+    return n
+  else
+    return fib(n - 1) + fib(n - 2)
+  end
+end
+
+# Time complexity: O(n), where n isnthe length of the input
+# Space complexity: O(1) because additional space in memory is not being consumed 
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  # raise NotImplementedError, "Method not implemented"
+  if s.nil? || s.length == 0 || s.length == 1
+    return s
+  else 
+    return s[-1] + reverse(s[0...-1])
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n), where n isnthe length of the input
+# Space complexity: O(1) because additional space in memory is not being consumed
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+    # raise NotImplementedError, "Method not implemented"
+    if s.nil? || s.length == 0 || s.length == 1
+      return s
+    else
+      return s[-1] + reverse(s[0...-1]) 
```

Comment:
This is not in place since you create a new string. 

Think about making a helper which takes the string, a left index and right index.  The base case is when `left_index >= right_index`  Each recursive call you swap the characters at `left_index` or `right_index` and then make a recursive call with the string, and `left_index + 1` and `right_index -1`.